### PR TITLE
scripts: added temp gpg key script

### LINKED ISSUE #2766: Update documentation on GPG and create a temp solution for quick start
### Proposed feature

Apps has a hard requirement on being able to encrypt secrets with GPG, but to simplify the setup for new Platform Administrators we should:
- first have better documentation to support the creation of GPG keys, and setup of SOPS, and 
- second have a fallback to support setting up a temporary GPG home and key for local testing

### Proposed alternatives

The first should be done on the public docs, as then we can link to it from any repo.

The second should be done here in Apps, as the prime use-case will be for local-clusters and Apps combo.

### Additional context

_No response_

### Definition of done

- [ ] Public docs for setting up GPG and SOPS
- [ ] Scripts and helpers to run a temporary GPG home for local testing

### PR DESCRIPTION
**Pull request by shafi-elastisys**
> [!warning]
> **This is a public repository, ensure not to disclose:**
>
> - [ ] personal data beyond what is necessary for interacting with this pull request, nor
> - [ ] business confidential information, such as customer names.

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [x] kind/feature       
- [ ] kind/improvement   
- [ ] kind/deprecation   
- [ ] kind/documentation 
- [ ] kind/clean-up      
- [ ] kind/bug           
- [ ] kind/other         

_Optional_: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [ ] kind/admin-change   
- [ ] kind/dev-change     
- [ ] kind/security       
- [ ] [kind/adr](set-me\) 

### What does this PR do / why do we need this PR?

This PR introduces script to which helps to setup a temporary GPG for local testing. 

* Added `bin/setup-local-gpg` which generaes a GPG Key non-interactively and exports necessary environment variables for SOPS to use. 
* Updated `bats.lib.bash` added helper functions. This allows test to invoke the script during the setup phase. 

- Fixes #2766 

#### Checklist

- [ ] Proper commit message prefix on all commits
    
- Change checks:
    - [x] The change is transparent
    - [ ] The change is disruptive
    - [ ] The change requires no migration steps
    - [ ] The change requires migration steps
    - [ ] The change updates CRDs
    - [ ] The change updates the config _and_ the schema
- Documentation checks:
    - [ ] The [public documentation](https://github.com/elastisys/welkin) required no updates
    - [ ] The [public documentation](https://github.com/elastisys/welkin) required an update - [link to change](set-me\)
- Metrics checks:
    - [ ] The metrics are still exposed and present in Grafana after the change
    - [ ] The metrics names didn't change (Grafana dashboards and Prometheus alerts required no updates)
    - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts required an update)
- Logs checks:
    - [ ] The logs do not show any errors after the change
- PodSecurityPolicy checks:
    - [ ] Any changed Pod is covered by Kubernetes Pod Security Standards
    - [ ] Any changed Pod is covered by Gatekeeper Pod Security Policies
    - [ ] The change does not cause any Pods to be blocked by Pod Security Standards or Policies
- NetworkPolicy checks:
    - [ ] Any changed Pod is covered by Network Policies
    - [ ] The change does not cause any dropped packets in the NetworkPolicy Dashboard
- Audit checks:
    - [ ] The change does not cause any unnecessary Kubernetes audit events
    - [ ] The change requires changes to Kubernetes audit policy
- Falco checks:
    - [ ] The change does not cause any alerts to be generated by Falco
- Bug checks:
    - [ ] The bug fix is covered by regression tests

### CODE REVIEW

File: tests/bats.lib.bash [hunk 1]
```diff
@@ -162,6 +162,22 @@ with_kubeconfig() {
...
+
+  eval "$(setup-local-gpg)"
+
+  if [[ -z "${CK8S_PGP_FP:-}" ]]; then
+    fail "Failed to generate temporary GPG key"
+  fi
+
+  log.trace "Temporary GPG initialized. Fingerprint: $CK8S_PGP_FP"
+}
+
```
Comment (aarnq) [inline tests/bats.lib.bash]: Is this needed? We already have a function for it already in `common/bats/gpg.sh` that also takes care of an edge case in which the GPG agent isn't available in the first tries.

File: scripts/local-cluster.sh [hunk 2]
```diff
@@ -304,8 +304,27 @@ config() {
...
 
+  local env_file="${CK8S_CONFIG_PATH}/.temp_gpg_env"
   if [[ -z "${CK8S_PGP_FP:-}" ]]; then
-    log.fatal "CK8S_PGP_FP is unset"
+    if [[ -f "${env_file}" ]]; then
+      # shellcheck disable=SC1090
+      source "${env_file}"
+      log.info "CK8S_PGP_FP was unset. Reusing temporary GPG environment from ${env_file}"
+      log.info "To persist this in your current terminal, run:"
+      log.info "  source ${env_file}"
```
Comment (aarnq) [inline scripts/local-cluster.sh]: Since it won't be too persistent if the temp gnupg home dir is cleared maybe we should make it seem less permanent 😄 same on the other print.
```suggestion
      log.info "To load this in your current terminal, run:"
      log.info "  source ${env_file}"
```

File: scripts/local-cluster.sh [hunk 3]
```diff
@@ -304,8 +304,27 @@ config() {
     log.fatal "CK8S_CONFIG_PATH is unset"
   fi
 
+  local env_file="${CK8S_CONFIG_PATH}/.temp_gpg_env"
```
Comment (aarnq) [inline scripts/local-cluster.sh]: We have a preference to use kebab case for files
```suggestion
  local env_file="${CK8S_CONFIG_PATH}/.temp-gpg-env"
```

File: scripts/local-cluster.sh [hunk 4]
```diff
@@ -353,6 +372,7 @@ config() {
   fi
 
   "${ROOT}/bin/ck8s" init both
+
```
Comment (aarnq) [inline scripts/local-cluster.sh]:
```suggestion
"${ROOT}/bin/ck8s" init both
```

File: scripts/local-cluster.sh [hunk 5]
```diff
@@ -304,8 +304,27 @@ config() {
...
   if [[ -z "${CK8S_PGP_FP:-}" ]]; then
-    log.fatal "CK8S_PGP_FP is unset"
+    if [[ -f "${env_file}" ]]; then
+      # shellcheck disable=SC1090
+      source "${env_file}"
+      log.info "CK8S_PGP_FP was unset. Reusing temporary GPG environment from ${env_file}"
+      log.info "To persist this in your current terminal, run:"
+      log.info "  source ${env_file}"
+    elif [[ -x "${ROOT}/bin/setup-local-gpg" ]]; then
+      eval "$("${ROOT}/bin/setup-local-gpg")"
```
Comment (simonklb) [inline scripts/local-cluster.sh]: 🚫 No eval!

Comment (shafi-elastisys) [reply scripts/local-cluster.sh]: Moved to source instead of eval.

File: bin/setup-local-gpg.bash [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+%commit
+%echo done
+EOF
+
+gpg --batch --generate-key "${GNUPGHOME}/gpg-batch" >/dev/null 2>&1
+
+# SOPS needs this fingerprint to know which key to use
+FINGERPRINT=$(gpg --list-secret-keys --keyid-format LONG | grep sec | awk '{print $2}' | cut -d'/' -f2)
+
+export GNUPGHOME="${GNUPGHOME}"
```
Comment (simonklb) [inline bin/setup-local-gpg.bash]: Already exported earlier in the script.

```suggestion
```

File: bin/setup-local-gpg.bash [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+export CK8S_PGP_FP="${FINGERPRINT}"
```
Comment (simonklb) [inline bin/setup-local-gpg.bash]:
```suggestion
export GNUPGHOME="${GNUPGHOME}"
# SOPS needs this fingerprint to know which key to use
export CK8S_PGP_FP="$(gpg --list-secret-keys --keyid-format LONG | grep sec | awk '{print $2}' | cut -d'/' -f2)"
```

File: scripts/local-cluster.sh (hunk 4 above)
Comment (simonklb) [inline scripts/local-cluster.sh]:
```suggestion
```

File: bin/setup-local-gpg.bash (hunk 7 above)
Comment (shafi-elastisys) [reply bin/setup-local-gpg.bash]: This is giving shell check fail as it is not recommended to avoid masking return values.

Comment (simonklb) [reply bin/setup-local-gpg.bash]: Ah right, my bad! The only problem I have with the current solution is that when you source it you will also get the `$FINGERPRINT` variable in your shell. So in that case, can you do something like this instead?
```
CK8S_PGP_FP=$(gpg --list-secret-keys --keyid-format LONG | grep sec | awk '{print $2}' | cut -d'/' -f2)
export CK8S_PGP_FP
```

Comment (shafi-elastisys) [reply bin/setup-local-gpg.bash]: Done with latest commit.

File: scripts/local-cluster.sh [hunk 8]
```diff
@@ -541,6 +541,147 @@ delete() {
   fi
 }
 
+status() {
+  local target_cluster="${1:-}"
+
+  local index_file="${CK8S_CONFIG_PATH}/cluster-index.yaml"
```
Comment (aarnq) [inline scripts/local-cluster.sh]: Why is this included here? Can you rebase and sort your prs by merging them into each other based on how they depend so we can have some clarity?

Comment (shafi-elastisys) [reply scripts/local-cluster.sh]: Sorry my bad. I wrongly commited into this branch. Working in many changes in different branches cause this confusion.

File: scripts/local-cluster.sh (hunk 2 above)
Comment (aarnq) [reply scripts/local-cluster.sh]: @shafi-elastisys you missed the first print.

File: scripts/local-cluster.sh (hunk 8 above)
Comment (shafi-elastisys) [reply scripts/local-cluster.sh]: Now it's fixed.

File: bin/setup-local-gpg.bash [hunk 9]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+GNUPGHOME="$(mktemp -d)"
+export GNUPGHOME
```
Comment (simonklb) [inline bin/setup-local-gpg.bash:5]: Hang on, won't this stop working if you reboot your computer?

Comment (aarnq) [reply bin/setup-local-gpg.bash:5]: Yes, hence the warnings in local-cluster.sh

Comment (simonklb) [reply bin/setup-local-gpg.bash:5]: Alright if this is acceptable then don't mind me. :smile:
